### PR TITLE
GFORMS-1714: Add top level expression to ATL so that aggregate expressions can be used

### PR DESCRIPTION
**Pull request by lxol**

### CODE REVIEW

File: test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala [hunk 1]
```diff
@@ -357,9 +357,9 @@ class EvaluationResultsSpec extends Spec with TableDrivenPropertyChecks {
   it should "evaluate form component reference from outside, when they are in indexed pages (repeating page, ATL)" in {
 
     val table = Table(
-      ("typeInfo", "recData", "evaluationContext", "expectedResult", "scenario"),
+      ("typeInfo", "recData", "evaluationContext", "expectedResult", "exprMap", "scenario"),
```
[the remark this answers is not in the document]
Comment (lxol) [reply test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala:360]: they are not longer working with empty `exprMap` because when we evaluate to `ListResult` we use only `exprMap` but previously we got it from `recData`

[the remark this answers is not in the document]
Comment (josef-vlach) [reply test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala:360]: Ok, that make sense

File: test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala [hunk 2]
```diff
@@ -357,9 +357,9 @@ class EvaluationResultsSpec extends Spec with TableDrivenPropertyChecks {
...
       (
-        TypeInfo(FormCtx(FormComponentId("addToListNumField")), StaticTypeData(ExprType.number, None)),
+        TypeInfo(Sum(FormCtx(FormComponentId("addToListNumField"))), StaticTypeData(ExprType.number, None)),
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala:362]: Why change to `Sum`?

File: test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala [hunk 3]
```diff
@@ -434,20 +446,34 @@ class EvaluationResultsSpec extends Spec with TableDrivenPropertyChecks {
       VariadicFormData.create(
         (toModelComponentId("1_addToListQuestion"), VariadicValue.One("0")),
         (toModelComponentId("2_addToListQuestion"), VariadicValue.One("1")),
+        (toModelComponentId("3_addToListQuestion"), VariadicValue.One("3")),
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala:449]: Why this change? `addToListQuestion` can never have answer `3` it is yes/no question.

File: test/uk/gov/hmrc/gform/models/FormModelSpec.scala [hunk 4]
```diff
@@ -807,6 +807,50 @@ class FormModelSpec extends AnyFlatSpecLike with Matchers with FormModelSupport
     }
   }
 
+  it should "576" in {
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/models/FormModelSpec.scala:810]: 576 what?

File: test/uk/gov/hmrc/gform/eval/ExpressionResultSpec.scala [hunk 5]
```diff
@@ -267,4 +267,32 @@ class ExpressionResultSpec extends FunSuite {
...
+    val nr2 = NumberResult(2)
+    val nr3 = NumberResult(3)
+    val lr12 = ListResult(List(nr1, nr2))
+    val lr13 = ListResult(List(nr1, nr3))
+
+    assertEquals(lr13 + lr12, ListResult(List(NumberResult(2), NumberResult(5))))
+    assertEquals(lr13 - lr12, ListResult(List(NumberResult(0), NumberResult(1))))
+    assertEquals(lr13 * lr12, ListResult(List(NumberResult(1), NumberResult(6))))
+    assertEquals(lr13 / lr12, ListResult(List(NumberResult(1), NumberResult(1.5))))
+  }
```
Comment (josef-vlach) [inline test/uk/gov/hmrc/gform/eval/ExpressionResultSpec.scala:297]: It would be great to cover interaction with `Hidden` and perhaps other cases too.

File: test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala (hunk 3 above)
Comment (lxol) [reply test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala:449]: some leftover from my experimentation, I will revert back

File: test/uk/gov/hmrc/gform/models/FormModelSpec.scala (hunk 4 above)
Comment (lxol) [reply test/uk/gov/hmrc/gform/models/FormModelSpec.scala:810]: oops,

File: test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala (hunk 2 above)
Comment (lxol) [reply test/uk/gov/hmrc/gform/eval/EvaluationResultsSpec.scala:362]: I think I replaced existing test instead of adding a new one. I will fix it.

File: test/uk/gov/hmrc/gform/eval/ExpressionResultSpec.scala (hunk 5 above)
Comment (lxol) [reply test/uk/gov/hmrc/gform/eval/ExpressionResultSpec.scala:320]: I tried to emulate ListRestult behavior on Hidden to how it works with NumberResults
```
Hidden + nr1
// res42: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(1)
Hidden - nr1
// res43: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(1)
Hidden * nr1
// res44: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(0)
Hidden / nr1
// res45: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(0)


nr1 + Hidden
// res46: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(1)
nr1 - Hidden
// res47: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(1)
nr1 * Hidden
// res48: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(0)
nr1 / Hidden
// res49: uk.gov.hmrc.gform.eval.ExpressionResult = NumberResult(0)
```

and this is how it works with lr12 = ListResult(List(NumberResult(1), NumberResult(2))))
```
Hidden + lr12
// res50: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(1), NumberResult(2)))
Hidden - lr12
// res51: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(1), NumberResult(2)))
Hidden * lr12
// res52: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(0), NumberResult(0)))
Hidden / lr12
// res53: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(0), NumberResult(0)))


lr12 + Hidden
// res54: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(1), NumberResult(2)))
lr12 - Hidden
// res55: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(1), NumberResult(2)))
lr12 * Hidden
// res56: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(0), NumberResult(0)))
lr12 / Hidden
// res57: uk.gov.hmrc.gform.eval.ExpressionResult = ListResult(List(NumberResult(0), NumberResult(0)))
```

**Review by josef-vlach — APPROVED** [review]
